lib/cstruct.ml

```diff
@@ -370,6 +370,35 @@ let of_bytes ?allocator buf =
     blit_from_bytes buf 0 c 0 buflen;
     set_len c buflen
 
+let string_fold ~f ~z str =
+  let st = ref z in
+  ( String.iter (fun c -> st := f !st c) str  ; !st )
+
+let of_hex str =
+  let hexdigit = function
+    | 'a' .. 'f' as x -> int_of_char x - 87
+    | 'A' .. 'F' as x -> int_of_char x - 55
+    | '0' .. '9' as x -> int_of_char x - 48
+    | x               -> Format.ksprintf invalid_arg "of_hex: `%c'" x
+  in
+  let whitespace = function
+    | ' ' | '\t' | '\r' | '\n' -> true
+    | _                        -> false
+  in
+  match
+    string_fold
+      ~f:(fun (cs, i, acc) -> function
+          | char when whitespace char -> (cs, i, acc)
+          | char ->
+            match (acc, hexdigit char) with
+            | (None  , x) -> (cs, i, Some (x lsl 4))
+            | (Some y, x) -> set_uint8 cs i (x lor y) ; (cs, succ i, None))
+      ~z:(create_unsafe (String.length str lsr 1), 0, None)
+      str
+  with
+  | (_ , _, Some _) -> Format.ksprintf invalid_arg "of_hex: dangling nibble"
```

Comment:
I had to look up `nibble`. It may be that I'm just not very educated, but I feel like `of_hex: uneven number of hex characters` would be easier to understand.

Reply:
removed any mention of `nibble`